File: packages/veil-sdk-dart/lib/src/discovery_parse.dart
import "dart:convert";

import "identity/contact_bundle.dart";
import "vps_profile.dart";

class DiscoveryParseResult {
  final List<String> wsEndpoints;
  final List<String> quicEndpoints;
  final List<String> peers;
  final List<String> tags;
  final String? quicCertHex;
  final String? quicCertB64;
  final ContactBundle? contactBundle;
  final bool isVpsProfile;

  const DiscoveryParseResult({
    required this.wsEndpoints,
    required this.quicEndpoints,
    required this.peers,
    required this.tags,
    this.quicCertHex,
    this.quicCertB64,
    this.contactBundle,
    this.isVpsProfile = false,
  });
}

DiscoveryParseResult? parseDiscoveryInput(String value) {
  final raw = value.trim();
  if (raw.isEmpty) return null;
  if (raw.startsWith("veil://contact")) {
    try {
      final bundle = ContactBundle.fromQrString(raw);
      return DiscoveryParseResult(
        wsEndpoints: const [],
        quicEndpoints: const [],
        peers: const [],
        tags: const [],
        contactBundle: bundle,
        isVpsProfile: false,
      );
    } catch (_) {
      return null;
    }
  }
  if (raw.startsWith("veil://") || raw.startsWith("veil:vps:") || raw.startsWith("vps:")) {
    final uri = _normalizeVpsUri(raw);
    if (uri == null) return null;
    final ws = uri.queryParametersAll["ws"] ?? const [];
    final peers = uri.queryParametersAll["peer"] ?? const [];
    final tags = uri.queryParametersAll["tag"] ?? const [];
    final quic = uri.queryParametersAll["quic"] ?? const [];
    final cert = uri.queryParameters["cert"];
    final certB64 = uri.queryParameters["certb64"];
    return DiscoveryParseResult(
      wsEndpoints: ws,
      quicEndpoints: quic,
      peers: peers,
      tags: tags,
      quicCertHex: cert,
      quicCertB64: certB64,
      contactBundle: null,
      isVpsProfile: true,
    );
  }
  if (raw.startsWith("http://") || raw.startsWith("https://")) {
    final uri = Uri.tryParse(raw);
    if (uri != null && (uri.path.isEmpty || uri.path == "/" || uri.path.endsWith("/config.js"))) {
      return DiscoveryParseResult(
        wsEndpoints: const [],
        quicEndpoints: const [],
        peers: const [],
        tags: const [],
        contactBundle: null,
        isVpsProfile: true,
      );
    }
  }
  if (raw.startsWith("ws://") || raw.startsWith("wss://")) {
    return DiscoveryParseResult(
      wsEndpoints: [raw],
      quicEndpoints: const [],
      peers: const [],
      tags: const [],
      contactBundle: null,
      isVpsProfile: false,
    );
  }
  if (raw.startsWith("quic://")) {
    return DiscoveryParseResult(
      wsEndpoints: const [],
      quicEndpoints: [raw],
      peers: const [],
      tags: const [],
      contactBundle: null,
      isVpsProfile: false,
    );
  }
  if (raw.startsWith("peer:")) {
    return DiscoveryParseResult(
      wsEndpoints: const [],
      quicEndpoints: const [],
      peers: [raw.substring(5)],
      tags: const [],
      contactBundle: null,
      isVpsProfile: false,
    );
  }
  if (raw.startsWith("tag:")) {
    return DiscoveryParseResult(
      wsEndpoints: const [],
      quicEndpoints: const [],
      peers: const [],
      tags: [raw.substring(4)],
      contactBundle: null,
      isVpsProfile: false,
    );
  }
  final hex = raw.toLowerCase().replaceAll(RegExp(r"[^0-9a-f]"), "");
  if (hex.length == 64) {
    return DiscoveryParseResult(
      wsEndpoints: const [],
      quicEndpoints: const [],
      peers: const [],
      tags: [hex],
      contactBundle: null,
      isVpsProfile: false,
    );
  }
  return null;
}

VpsProfile? parseVpsProfileFromConfig(String host, String body) {
  return VpsProfile.parseConfigJs(host, body);
}

Uri? _normalizeVpsUri(String raw) {
  final lower = raw.toLowerCase();
  if (lower.startsWith("veil://")) {
    return Uri.tryParse(raw);
  }
  if (lower.startsWith("vps:")) {
    return Uri.tryParse("veil://${raw.substring(4)}");
  }
  if (lower.startsWith("veil:vps:")) {
    return Uri.tryParse("veil://${raw.substring(9)}");
  }
  return null;
}

String? tryDecodeVpsProfileFromConfigJs(String host, String body) {
  final profile = VpsProfile.parseConfigJs(host, body);
  if (profile == null) return null;
  return jsonEncode({
    "host": profile.host,
    "ws": profile.wsUrl,
    if (profile.quicEndpoint != null) "quic": profile.quicEndpoint,
    if (profile.quicCertB64 != null) "certb64": profile.quicCertB64,
  });
}

File: packages/veil-sdk-js/src/discovery_parse.ts
import {
  contactBundleFromQr,
  type ContactBundle,
} from "./identity";

export type DiscoveryParseResult = {
  wsEndpoints: string[];
  quicEndpoints: string[];
  peers: string[];
  tags: string[];
  quicCertHex?: string;
  quicCertB64?: string;
  contactBundle?: ContactBundle;
  isVpsProfile: boolean;
};

export function parseDiscoveryInput(value: string): DiscoveryParseResult | null {
  const raw = value.trim();
  if (!raw) return null;
  if (raw.startsWith("veil://contact")) {
    try {
      const bundle = contactBundleFromQr(raw);
      return {
        wsEndpoints: [],
        quicEndpoints: [],
        peers: [],
        tags: [],
        contactBundle: bundle,
        isVpsProfile: false,
      };
    } catch {
      return null;
    }
  }
  if (raw.startsWith("veil://") || raw.startsWith("veil:vps:") || raw.startsWith("vps:")) {
    const uri = normalizeVpsUri(raw);
    if (!uri) return null;
    const ws = uri.searchParams.getAll("ws");
    const peers = uri.searchParams.getAll("peer");
    const tags = uri.searchParams.getAll("tag");
    const quic = uri.searchParams.getAll("quic");
    const cert = uri.searchParams.get("cert") ?? undefined;
    const certb64 = uri.searchParams.get("certb64") ?? undefined;
    return {
      wsEndpoints: ws,
      quicEndpoints: quic,
      peers,
      tags,
      quicCertHex: cert,
      quicCertB64: certb64,
      isVpsProfile: true,
    };
  }
  if (raw.startsWith("http://") || raw.startsWith("https://")) {
    try {
      const url = new URL(raw);
      if (!url.pathname || url.pathname === "/" || url.pathname.endsWith("/config.js")) {
        return {
          wsEndpoints: [],
          quicEndpoints: [],
          peers: [],
          tags: [],
          isVpsProfile: true,
        };
      }
    } catch {
      return null;
    }
  }
  if (raw.startsWith("ws://") || raw.startsWith("wss://")) {
    return {
      wsEndpoints: [raw],
      quicEndpoints: [],
      peers: [],
      tags: [],
      isVpsProfile: false,
    };
  }
  if (raw.startsWith("quic://")) {
    return {
      wsEndpoints: [],
      quicEndpoints: [raw],
      peers: [],
      tags: [],
      isVpsProfile: false,
    };
  }
  if (raw.startsWith("peer:")) {
    return {
      wsEndpoints: [],
      quicEndpoints: [],
      peers: [raw.slice(5)],
      tags: [],
      isVpsProfile: false,
    };
  }
  if (raw.startsWith("tag:")) {
    return {
      wsEndpoints: [],
      quicEndpoints: [],
      peers: [],
      tags: [raw.slice(4)],
      isVpsProfile: false,
    };
  }
  const hex = raw.toLowerCase().replace(/[^0-9a-f]/g, "");
  if (hex.length === 64) {
    return {
      wsEndpoints: [],
      quicEndpoints: [],
      peers: [],
      tags: [hex],
      isVpsProfile: false,
    };
  }
  return null;
}

function normalizeVpsUri(raw: string): URL | null {
  const lower = raw.toLowerCase();
  if (lower.startsWith("veil://")) {
    return new URL(raw);
  }
  if (lower.startsWith("vps:")) {
    return new URL(`veil://${raw.slice(4)}`);
  }
  if (lower.startsWith("veil:vps:")) {
    return new URL(`veil://${raw.slice(9)}`);
  }
  return null;
}
